clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonLoadBalancerInstanceStateCachingAgent.groovy

```diff
@@ -98,39 +100,60 @@ class AmazonLoadBalancerInstanceStateCachingAgent implements CachingAgent, Healt
+    Collection<CacheData> instanceRels = new ArrayList<>()
 
-    Collection<CacheData> lbHealths = []
-    Collection<CacheData> instances = []
     for (loadBalancerKey in loadBalancerKeys) {
       try {
         Map<String, String> idObj = Keys.parse(loadBalancerKey)
-        def lbName = idObj.loadBalancer
+        String lbName = idObj.loadBalancer
         if (idObj.loadBalancerType && idObj.loadBalancerType != 'classic')
           continue
 
-        def result = loadBalancing.describeInstanceHealth(new DescribeInstanceHealthRequest(lbName))
-        def loadBalancerInstances = []
+        List<LoadBalancerInstance> loadBalancerInstances = new ArrayList<>()
+        DescribeInstanceHealthResult result = loadBalancing
+          .describeInstanceHealth(new DescribeInstanceHealthRequest(lbName))
+
         for (instanceState in result.instanceStates) {
-          def loadBalancerInstance = new LoadBalancerInstance(instanceState.instanceId, instanceState.state, instanceState.reasonCode, instanceState.description)
+          LoadBalancerInstance loadBalancerInstance = new LoadBalancerInstance(
+            instanceState.instanceId,
+            instanceState.state,
+            instanceState.reasonCode,
+            instanceState.description)
           loadBalancerInstances << loadBalancerInstance
         }
-        def loadBalancerInstanceState = new LoadBalancerInstanceState(name: lbName, instances: loadBalancerInstances)
-        def ilbs = InstanceLoadBalancers.fromLoadBalancerInstanceState([loadBalancerInstanceState])
+
+        LoadBalancerInstanceState loadBalancerInstanceState = new LoadBalancerInstanceState(
+          name: lbName,
+          instances: loadBalancerInstances)
+        List<InstanceLoadBalancers> ilbs = InstanceLoadBalancers
+          .fromLoadBalancerInstanceState([loadBalancerInstanceState])
+        Collection<String> instanceIds = ilbs.collect {
+          Keys.getInstanceKey(it.instanceId, account.name, region)
+        }
+        Map<String, CacheData> instances = providerCache
+          .getAll(INSTANCES.ns, instanceIds, RelationshipCacheFilter.none())
+          .collectEntries { [(it.id): it] }
 
         for (InstanceLoadBalancers ilb in ilbs) {
           String instanceId = Keys.getInstanceKey(ilb.instanceId, account.name, region)
           String healthId = Keys.getInstanceHealthKey(ilb.instanceId, account.name, region, healthId)
-
           Map<String, Object> attributes = objectMapper.convertValue(ilb, ATTRIBUTES)
-          if (idObj.containsKey("application")) {
-            attributes.put("application", idObj.get("application"))
+          Map<String, Collection<String>> relationships = [(INSTANCES.ns): [instanceId]]
+
+          if (instances[instanceId] != null) {
+            String application = instances[instanceId].attributes.get("application")
+            if (application != null) {
+              attributes.put("application", application)
```

Comment:
This ensures that health objects stored by `AmazonLoadBalancerInstanceStateCachingAgent` are indexed by the application of instances the objects relate to, instead of the application the load balancer itself. This was already done by `EddaLoadBalancerCachingAgent` which runs instead of `AmazonLoadBalancerInstanceStateCachingAgent` when edda is enabled.